### PR TITLE
Add trace id to logs

### LINKED ISSUE #238: Add a trace id to logs
### Problem

We have logs but with multiple request at a time it can be difficult to correlate logs

### Solution

Add a request/correlation id that follows activity through the stack

### Alternatives

_No response_

### PR DESCRIPTION
**Pull request by georgi-l95**
**Description**:
Added a request id that follows activity through the stack.

**Related issue(s)**:

Fixes #238 

**Notes for reviewer**:
``` 
[2022-08-09 06:55:27.204 +0000] DEBUG (rpc-server/65064 on Georgis-MacBook-Pro-3.local): [Request ID: f61ff7bb-d6ea-40a3-b6d9-185b3b5ff60f] eth_getTransactionReceipt
[2022-08-09 06:55:27.204 +0000] TRACE (relay-eth/65064 on Georgis-MacBook-Pro-3.local): [Request ID: f61ff7bb-d6ea-40a3-b6d9-185b3b5ff60f] getTransactionReceipt(0xbe547272e48ffc1c0ee2d54c98f6ff4243e6f4e1b53c1872652f452cf2a77e4a)
[2022-08-09 06:55:27.226 +0000] DEBUG (mirror-node/65064 on Georgis-MacBook-Pro-3.local): [Request ID: f61ff7bb-d6ea-40a3-b6d9-185b3b5ff60f] [GET] contracts/results/0xbe547272e48ffc1c0ee2d54c98f6ff4243e6f4e1b53c1872652f452cf2a77e4a 200 22 ms
[2022-08-09 06:55:27.226 +0000] TRACE (relay-eth/65064 on Georgis-MacBook-Pro-3.local): [Request ID: f61ff7bb-d6ea-40a3-b6d9-185b3b5ff60f] receipt for 0xbe547272e48ffc1c0ee2d54c98f6ff4243e6f4e1b53c1872652f452cf2a77e4a found in block 0x22
[2022-08-09 06:55:27.226 +0000] INFO (rpc-server/65064 on Georgis-MacBook-Pro-3.local): [Request ID: f61ff7bb-d6ea-40a3-b6d9-185b3b5ff60f] [POST] eth_getTransactionReceipt: 200 22 ms 
[2022-08-09 06:55:27.234 +0000] DEBUG (rpc-server/65064 on Georgis-MacBook-Pro-3.local): [Request ID: e391e121-04f3-477f-b557-fa4f0d178d67] eth_chainId
[2022-08-09 06:55:27.234 +0000] TRACE (relay-eth/65064 on Georgis-MacBook-Pro-3.local): [Request ID: e391e121-04f3-477f-b557-fa4f0d178d67] chainId()
[2022-08-09 06:55:27.234 +0000] INFO (rpc-server/65064 on Georgis-MacBook-Pro-3.local): [Request ID: e391e121-04f3-477f-b557-fa4f0d178d67] [POST] eth_chainId: 200 0 ms 
[2022-08-09 06:55:27.237 +0000] DEBUG (rpc-server/65064 on Georgis-MacBook-Pro-3.local): [Request ID: 528ec90e-9d28-4824-b89c-61c72a05ab63] eth_gasPrice
[2022-08-09 06:55:27.237 +0000] TRACE (relay-eth/65064 on Georgis-MacBook-Pro-3.local): [Request ID: 528ec90e-9d28-4824-b89c-61c72a05ab63] gasPrice()
[2022-08-09 06:55:27.237 +0000] INFO (rpc-server/65064 on Georgis-MacBook-Pro-3.local): [Request ID: 528ec90e-9d28-4824-b89c-61c72a05ab63] [POST] eth_gasPrice: 200 0 ms 
[2022-08-09 06:55:27.237 +0000] DEBUG (rpc-server/65064 on Georgis-MacBook-Pro-3.local): [Request ID: 7834f00b-0ebc-4368-9bf5-0d104252eb04] eth_getBlockByNumber

```
**Checklist**

- [ ] Documented (Code comments, README, etc.)
- [ ] Tested (unit, integration, etc.)

### CODE REVIEW

File: packages/server/src/server.ts [hunk 1]
```diff
@@ -46,6 +47,8 @@ const cors = require('koa-cors');
 const app = new Koa();
 const rpc = koaJsonRpc();
 
+const REQUEST_ID_STRING = `Request ID`;
+const REQUEST_ID_MAX_VALUE = 100000000;
```
Comment (Nana-EC) [inline packages/server/src/server.ts:51]: I think this is an okay start.
In future with more request we'll want to explore this range or consider alpha numeric values to avoid collisions.
Knowing the time range of an in issue will likely avoid collision confusion so this isn't a biggie

File: packages/server/src/server.ts [hunk 2]
```diff
@@ -191,7 +207,7 @@ rpc.use('eth_blockNumber', async () => {
  * returns: Gas used - hex encoded integer
  */
 rpc.use('eth_estimateGas', async (params: any) => {
-  return logAndHandleResponse('eth_estimateGas', () => relay.eth().estimateGas(params?.[0], params?.[1]));
+  return logAndHandleResponse('eth_estimateGas', (requestId) => relay.eth().estimateGas(params?.[0], params?.[1],requestId));
```
Comment (Nana-EC) [inline packages/server/src/server.ts:210]: nit: here and instances below
```suggestion
  return logAndHandleResponse('eth_estimateGas', (requestId) => relay.eth().estimateGas(params?.[0], params?.[1], requestId));
```

File: packages/server/src/server.ts [hunk 3]
```diff
@@ -161,17 +166,28 @@ const logAndHandleResponse = async (methodName, methodFunction) => {
...
+  const id = crypto.randomInt(REQUEST_ID_MAX_VALUE);
+  return id;
+};
+
+
 /**
  * returns: false
  */
-rpc.use('net_listening', async () => {
+rpc.use('net_listening',async () => {
```
Comment (Nana-EC) [inline packages/server/src/server.ts:183]: nit: revert this change here and below
```suggestion
rpc.use('net_listening', async () => {
```

File: packages/relay/src/lib/clients/mirrorNodeClient.ts [hunk 4]
```diff
@@ -140,110 +140,113 @@ export class MirrorNodeClient {
...
+        this.logger.error(new Error(error.message), `${requestIdPrefix} [GET] ${path} ${effectiveStatusCode} status`);
         throw new MirrorNodeClientError(error.message, effectiveStatusCode);
     }
 
-    public async getAccountLatestTransactionByAddress(idOrAliasOrEvmAddress: string): Promise<object> {
+    public async getAccountLatestTransactionByAddress(idOrAliasOrEvmAddress: string, requestId?: number): Promise<object> {
         return this.request(`${MirrorNodeClient.GET_ACCOUNTS_ENDPOINT}${idOrAliasOrEvmAddress}?order=desc&limit=1`,
             MirrorNodeClient.GET_ACCOUNTS_ENDPOINT,
-            [400]);
+            [400], requestId);
```
Comment (Nana-EC) [inline packages/relay/src/lib/clients/mirrorNodeClient.ts:178]: nit: next line for conformance.
Here and below

File: packages/relay/src/lib/constants.ts [hunk 5]
```diff
@@ -53,4 +53,5 @@ export default {
     TX_DEFAULT_GAS: 400_000,
     TX_CREATE_EXTRA: 32_000,
     TX_DATA_ZERO_COST: 4,
+    REQUEST_ID_STRING: `Request ID`
```
Comment (Nana-EC) [inline packages/relay/src/lib/constants.ts:56]: If you add `: ` to this constant you save the need to add them in every other log line you've added them to.

```suggestion
    REQUEST_ID_STRING: `Request ID: `
```

File: packages/server/src/server.ts [hunk 6]
```diff
@@ -161,17 +166,28 @@ const logAndHandleResponse = async (methodName, methodFunction) => {
   }
 };
 
+/**
+ * Generates random trace id for requests.
+ * 
+ * returns: integer
+ */
+ const generateRequestId = () => {
+  const id = crypto.randomInt(REQUEST_ID_MAX_VALUE);
```
Comment (Nana-EC) [inline packages/server/src/server.ts:175]: nit: just return instead of storing and returning

**Review by Nana-EC — CHANGES_REQUESTED** [review]
Looking good.
A few nits.

This is a good foundation to make troubleshooting a little easier.
Might need improvement in the future but those are out of scope for now.

One thing to note is if there's a way for use to pass this back to the caller in the JSON RPC spec without breaking the schema.
In this way callers could easily to the correlation of request to stack trace.

File: packages/server/src/server.ts (hunk 1 above)
Comment (georgi-l95) [reply packages/server/src/server.ts:51]: We could use randomUUID from crypto module. This will generate something like this 
`[2022-08-05 21:24:52.640 +0000] INFO (rpc-server/3975 on Georgis-MacBook-Pro-2.local): [Request ID:48f98bb9-45a2-4206-a0be-75482b8abe53] [POST] eth_sendRawTransaction: 200 571 ms 
`

Comment (Nana-EC) [reply packages/server/src/server.ts:51]: Yes, much better.
Make it so :)

**Comment by georgi-l95** [comment]
> One thing to note is if there's a way for use to pass this back to the caller in the JSON RPC spec without breaking the schema. In this way callers could easily to the correlation of request to stack trace.

I don't think there is an easy way of doing this without breaking the schema. Nevertheless I'd suggest we explore adding this in a new separate issue.

File: packages/relay/src/lib/eth.ts [hunk 7]
```diff
@@ -125,19 +125,21 @@ export class EthImpl implements Eth {
...
-  async feeHistory(blockCount: number, newestBlock: string, rewardPercentiles: Array<number> | null) {
-    this.logger.trace(`feeHistory(blockCount=${blockCount}, newestBlock=${newestBlock}, rewardPercentiles=${rewardPercentiles})`);
+  async feeHistory(blockCount: number, newestBlock: string, rewardPercentiles: Array<number> | null, requestId?: number) {
+    const requestIdPrefix = requestId ? `[${constants.REQUEST_ID_STRING}${requestId}]` : '';
+    this.logger.trace(`${requestIdPrefix} feeHistory(blockCount=${blockCount}, newestBlock=${newestBlock}, rewardPercentiles=${rewardPercentiles})`);
     try {
-      const latestBlockNumber = await this.translateBlockTag('latest');
-      const newestBlockNumber = await this.translateBlockTag(newestBlock);
+      const latestBlockNumber = await this.translateBlockTag('latest',requestId);
+      const newestBlockNumber = await this.translateBlockTag(newestBlock,requestId);
```
Comment (Nana-EC) [inline packages/relay/src/lib/eth.ts:142]: nit:
```suggestion
      const newestBlockNumber = await this.translateBlockTag(newestBlock, requestId);
```

File: packages/relay/src/lib/eth.ts [hunk 8]
```diff
@@ -152,33 +154,33 @@ export class EthImpl implements Eth {
       let feeHistory: object | undefined = cache.get(constants.CACHE_KEY.FEE_HISTORY);
       if (!feeHistory) {
 
-        feeHistory = await this.getFeeHistory(blockCount, newestBlockNumber, latestBlockNumber, rewardPercentiles);
+        feeHistory = await this.getFeeHistory(blockCount, newestBlockNumber, latestBlockNumber, rewardPercentiles,requestId);
```
Comment (Nana-EC) [inline packages/relay/src/lib/eth.ts:157]: nit: 
```suggestion
        feeHistory = await this.getFeeHistory(blockCount, newestBlockNumber, latestBlockNumber, rewardPercentiles, requestId);
```

File: packages/relay/src/lib/eth.ts [hunk 9]
```diff
@@ -125,19 +125,21 @@ export class EthImpl implements Eth {
    * This method is implemented to always return an empty array. This is in alignment
    * with the behavior of Infura.
    */
-  accounts() {
-    this.logger.trace('accounts()');
+  accounts(requestId?: number) {
+    const requestIdPrefix = requestId ? `[${constants.REQUEST_ID_STRING}${requestId}]` : '';
```
Comment (Nana-EC) [inline packages/relay/src/lib/eth.ts:129]: You can move this to a private function so instead of repeating the logic you just call getRequestIdMessage(requestId)
```suggestion
    const requestIdPrefix = getRequestIdMessage(requestId);
```

where
```
  private getRequestIdMessage(requestId: string): string {
    return requestId ? `[${constants.REQUEST_ID_STRING}${requestId}]` : '';
  }
```

File: packages/relay/src/lib/precheck.ts [hunk 10]
```diff
@@ -134,36 +137,37 @@ export class Precheck {
     let tinybars;
 
     try {
-      const { account }: any = await this.mirrorNodeClient.getAccount(tx.from!);
-      tinybars = await this.sdkClient.getAccountBalanceInTinyBar(account, callerName);
+      const { account }: any = await this.mirrorNodeClient.getAccount(tx.from!,requestId);
```
Comment (Nana-EC) [inline packages/relay/src/lib/precheck.ts:140]: nit: 
```suggestion
      const { account }: any = await this.mirrorNodeClient.getAccount(tx.from!, requestId);
```

**Review by Nana-EC — COMMENTED** [review]
Functionally good.
Just nits to to be fixed and get the DCO github action passing and this is good to merge.

**Review by Nana-EC — APPROVED** [review]
LG
